### PR TITLE
[draft] support ATOM plugin for qwen3.5 DPxTPx/DPxEPx

### PR DESCRIPTION
**Pull request by zovonoir**
## Summary

Add Qwen3.5 SGLang plugin support for DP-attention layouts, including DP+TP and DP+EP execution.

This change makes the Qwen3.5-specific language-model adapter use the same `SGLangPluginRuntime` lifecycle as the generic ATOM SGLang wrapper, and fixes the nested GDN context handling that previously caused rank-divergent collectives.

## Problem

Qwen3.5 uses a dedicated SGLang language-model adapter instead of the generic ATOM causal-LM wrapper. As a result:

- idle DP ranks could pass zero-token tensors into AITER GEMM kernels (`M=0`);
- Qwen3.5-VL idle forwards could omit `positions` and `input_ids`;
- mRoPE positions shaped `[3, num_tokens]` were interpreted as three tokens;
- each rank could build a different physical DP token-count vector after idle batches were materialized;
- the nested GDN binder rebuilt `ForwardContext`/`DPMetadata` and issued a second CPU all-reduce only on ranks with valid GDN metadata, while idle ranks skipped it and later waited in a different collective.

The last issue produced a deterministic distributed deadlock: active ranks waited in GDN DP metadata synchronization while idle ranks progressed to the logits all-gather.

### Qwen3.5 SGLang adapter

- Wrap the Qwen3.5 language-model forward with `plugin_runtime_scope` and `SGLangPluginRuntime`.
- Build SGLang metadata from the runtime-normalized batch.
- Forward the normalized `input_ids`, `positions`, and `input_embeds` to the ATOM decoder.
- Trim physical dummy-token outputs back to the logical SGLang token count.

### SGLang forward-context bridge

- Materialize an idle batch from any available tensor (`positions`, `input_ids`, or `input_embeds`).
- Create fallback dummy positions/token IDs when Qwen3.5-VL does not provide them.
- Normalize every zero entry in the cross-DP token-count vector to one physical dummy token so all ranks use identical `DPMetadata`.
- Read the token dimension correctly for mRoPE `[3, num_tokens]` positions.
- Propagate SGLang's global forward mode into `dp_uniform_decode` so mixed EXTEND/IDLE steps use the variable-length DP path.

### GDN context handling

- Reuse the cross-rank-consistent `ForwardContext` and `DPMetadata` already created by `SGLangPluginRuntime`.
- Temporarily inject only GDN attention metadata and KV-cache data.
- Restore the previous metadata/cache on exit instead of resetting the complete context.
- Retain the previous standalone fallback when no outer context exists.

This prevents GDN from creating a second, rank-local DP synchronization point.

## Parallel layouts covered

For the full-DP SGLang configurations used here:

- `TPN`: attention/dense TP=N, expert TP=N.
- `DPNTPN`: attention DP=N with physical attention TP=1; experts use logical TP=N over the DP ranks.
- `DPNEPN`: attention DP=N with physical attention TP=1; experts use EP=N and TP=1. When MORI is available, expert tokens use MORI dispatch/combine.

## Performance results

Measured Qwen3.5-397B results on AMD Instinct MI308X GPUs with OSL=2048. TP8 and TP8DP8 use the separate concurrency settings shown below, so the comparison reflects the tested configurations rather than equal-load scaling.

| Layout | ISL/OSL | Concurrency | Prompts | Successful | Total throughput (tok/s) | Input throughput (tok/s) | Output throughput (tok/s) | Median TTFT (ms) | Median TPOT (ms) | Duration (s) |
|---|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|
| TP8 | 4096/2048 | 216 | 432 | 432 | 5451.97 | 3634.65 | 1817.32 | 37408.86 | 100.70 | 486.83 |
| TP8 | 6144/2048 | 160 | 320 | 320 | 5651.08 | 4238.31 | 1412.77 | 40645.01 | 93.52 | 463.88 |
| TP8 | 14336/2048 | 80 | 160 | 160 | 6784.86 | 5936.75 | 848.11 | 49087.91 | 70.38 | 386.37 |
| TP8 | 30720/2048 | 40 | 80 | 80 | 7087.62 | 6644.64 | 442.98 | 50416.43 | 65.79 | 369.86 |
| TP8 | 63488/2048 | 20 | 40 | 40 | 6747.36 | 6536.51 | 210.86 | 52134.43 | 69.40 | 388.51 |
| TP8 | 129024/2048 | 8 | 16 | 16 | 6547.94 | 6445.62 | 102.31 | 55508.87 | 51.22 | 320.28 |
| TP8DP8 | 4096/2048 | 608 | 1216 | 1216 | 6972.32 | 4648.21 | 2324.11 | 147548.01 | 127.51 | 1071.54 |
| TP8DP8 | 6144/2048 | 512 | 1024 | 1024 | 7301.85 | 5476.39 | 1825.46 | 165676.38 | 143.10 | 1148.83 |
| TP8DP8 | 14336/2048 | 256 | 512 | 512 | 7688.74 | 6727.65 | 961.09 | 118323.00 | 196.07 | 1091.03 |
| TP8DP8 | 30720/2048 | 128 | 256 | 256 | 7499.75 | 7031.01 | 468.73 | 112871.17 | 191.42 | 1118.52 |
| TP8DP8 | 63488/2048 | 64 | 128 | 128 | 7142.03 | 6918.84 | 223.19 | 172455.53 | 183.74 | 1174.54 |
| TP8DP8 | 129024/2048 | 32 | 64 | 64 | 5178.52 | 5097.60 | 80.91 | 213152.23 | 249.80 | 1619.89 |

Total throughput change for TP8DP8 relative to TP8 at the listed concurrency settings: +27.9%, +29.2%, +13.3%, +5.8%, +5.8%, and -20.9% from ISL 4096 through 129024, respectively.

### CODE REVIEW

File: atom/plugin/sglang/models/qwen3_5.py [hunk 1]
```diff
@@ -265,48 +268,65 @@ def embed_input_ids(self, input_ids: torch.Tensor) -> torch.Tensor:
...
+                        inputs_embeds=runtime.input_embeds,
                         input_deepstack_embeds=input_deepstack_embeds,
                     )
-            if isinstance(out, IntermediateTensors):
-                return PPProxyTensors(dict(out.tensors))
-            return out
+                out = runtime.trim_output(out)
+                if isinstance(out, IntermediateTensors):
+                    return PPProxyTensors(dict(out.tensors))
+                return out
```
[the remark this answers is not in the document]
Comment (zovonoir) [reply atom/plugin/sglang/models/qwen3_5.py:366]: Fixed in a61253b2. The runtime output adapter now slices `IntermediateTensors` as well as tensors/tuples, so an IDLE rank returns zero-token pipeline payloads to SGLang while preserving the container type.

File: atom/plugin/sglang/models/qwen3_5.py [hunk 2]
```diff
@@ -265,48 +268,65 @@ def embed_input_ids(self, input_ids: torch.Tensor) -> torch.Tensor:
...
+                plugin_runtime_scope(framework="sglang", atom_config=self.atom_config),
+                SGLangPluginRuntime(
+                    atom_config=self.atom_config,
+                    forward_batch=forward_batch,
+                    positions=positions,
+                    input_ids=input_ids,
+                    input_embeds=inputs_embeds,
+                    set_forward_context=True,
+                ) as runtime,
+            ):
```
[the remark this answers is not in the document]
Comment (zovonoir) [reply atom/plugin/sglang/models/qwen3_5.py:335]: Added focused regression coverage in a61253b2. The tests exercise IDLE-batch one-token materialization followed by zero-token `IntermediateTensors` trimming, and verify that the full per-DP token vector normalizes every idle rank from 0 to 1.

**Comment by zovonoir** [comment]
@valarLip Could you please take a look when you have a chance? The review comments have been addressed, the branch is conflict-free, and all current CI checks are green. Thanks!

**Comment by ZhiweiYan-96** [comment]
hi, @zovonoir  Please rebase the PR for resolving the conflict with TBO new changes.

**Comment by zovonoir** [comment]
@ZhiweiYan-96 Rebased onto the latest `main` (`5d771f71`) and resolved the TBO-related conflicts. The PR is conflict-free again.

I also addressed the remaining review suggestions in `d1697127`:
- preserve the outer attention metadata via a shallow copy when injecting GDN metadata;
- fall back to the local forward mode when `global_forward_mode` is unavailable;
- preserve Qwen-VL mRoPE dummy-position shape and handle optional `seq_lens` / `seq_lens_cpu`;
- align the Qwen wrapper runtime mock with the real context-manager API;
- add focused regression coverage for these paths.

Local Ruff/Black checks pass, and the independently runnable SGLang regression tests pass (`5 passed`). Full CI has been triggered by the updated branch.

**Review by valarLip — DISMISSED** [review]

**Review by valarLip — APPROVED** [review]
